frontend/src/components/NewRunParameters.tsx

```diff
@@ -39,23 +61,150 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
+  param: ApiParameter;
+  index: number;
+  handleParamChange: (index: number, value: string) => void;
+}
+
+interface EnhancedTextFieldState {
+  isEditorOpen: boolean;
+  isInJsonForm: boolean;
+  isJsonField: boolean;
+}
+
+class EnhancedTextField extends React.Component<EnhancedTextFieldProps, EnhancedTextFieldState> {
+
+  public static getDerivedStateFromProps(nextProps: EnhancedTextFieldProps, prevState: EnhancedTextFieldState): EnhancedTextFieldState {
+    let isJson = true;
+    try {
+      const displayValue = JSON.parse(nextProps.param.value || '');
+      // Nulls, booleans, strings, and numbers can all be parsed as JSON, but we don't care
+      // about rendering. Note that `typeOf null` returns 'object'
+      if (displayValue === null || typeof displayValue !== 'object') {
+        throw new Error(
+          'Parsed JSON was neither an array nor an object. Using default renderer',
+        );
+      }
+    } catch (err) {
+      isJson = false;
+    }
+    return {
+      isEditorOpen: prevState.isEditorOpen,
+      isInJsonForm: isJson,
+      isJsonField: prevState.isJsonField || isJson
+    };
+  }
+
+  constructor(props: any) {
+    super(props);
+
+    this.state = {
+      isEditorOpen: false,
+      isInJsonForm: false,
+      isJsonField: false
+    };
+  }
+
+  public render(): JSX.Element | null {
+    const {param, index, handleParamChange} = this.props;
+
+    if (this.state.isJsonField){
+      return (
+        <>
+          <TextField
+            id={`newRunPipelineParam${index}`}
+            disabled={this.state.isEditorOpen}
+            variant='outlined'
+            label={param.name}
+            value={param.value || ''}
+            onChange={ev => handleParamChange(index, ev.target.value || '')}
+            style={{ maxWidth: 600 }}
+            className={commonCss.textField}
+            InputProps={{
+              classes: { disabled: css.nonEditableInput },
+              endAdornment: (
+                <InputAdornment position='end'>
+                  <Button
+                    color='secondary'
+                    id='chooseExperimentBtn'
+                    onClick={() => {
+                      if (this.state.isInJsonForm) {
+                        const displayValue = JSON.parse(param.value || '');
+                        if (this.state.isEditorOpen) {
+                          handleParamChange(index, JSON.stringify(displayValue) || '');
+                        } else {
+                          handleParamChange(
+                            index,
+                            JSON.stringify(displayValue, null, 2) || '',
+                          );
+                        }
+                      }
+                      this.setState({
+                        isEditorOpen: !this.state.isEditorOpen,
+                        isInJsonForm: this.state.isInJsonForm,
+                        isJsonField: this.state.isJsonField
+                      });
```

Comment:
setState receives a patch and updates state based on it, there's no need to specify isInJsonForm and isJsonField if they don't need to change.